agent/src/beerocks/slave/tasks/task_pool.cpp

```diff
@@ -0,0 +1,53 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "task_pool.h"
+
+using namespace beerocks;
+
+bool TaskPool::add_task(std::shared_ptr<Task> new_task)
```

Comment:
```suggestion
bool TaskPool::add_task(Task& new_task)
```

Reply:
Sorry, but that is a terrible idea because it forces the caller to keep the task object. Not something we want.

Reply:
There are many ways to express ownership in C++. Ownership is _important_ in code structure as it sets lifetimes, responsibilities and overall structure. Declaring `std::shared_ptr` is our "easy" way to forget about ownership and lifetime of objects. It is like that we are having a garbage collector to keep the memory and lifetime in a good shape.
So the question raised based on your comment is: who is responsible to keep the Task object?
If it is someone else, outside the `TaskPool`, then fine, just request a reference, someone else will have the real object.
If it is the `TaskPool` that keeps the task objects (sounds reasonable to me), then state it clearly:
```suggestion
bool TaskPool::add_task(Task&& new_task);
```
Now the ownership is _moved_ to the TaskPool.
If you could also handle a nullptr (which you can't, but for the example only) -
```suggestion
bool TaskPool::add_task(std::unique_ptr<Task> new_task)
```
Also I am not against _copying_ the task internally into the `TaskPool` -
```suggestion
bool TaskPool::add_task(Task new_task) {
m_task_pool.push_back(std::move(new_task));
```